amm/drink-tests/src/stable_swap_tests/tests_add_remove_lp.rs

```diff
@@ -0,0 +1,400 @@
+        CHARLIE,
+        vec![500 * ONE_DAI, 500 * ONE_USDT, 500 * ONE_USDC],
+        BOB,
+    );
+
+    // add more liquidity with balanced tokens (charlie)
+    _ = stable_swap::add_liquidity(
+        &mut session,
+        stable_swap,
+        CHARLIE,
+        1,
+        vec![500 * ONE_DAI, 500 * ONE_USDT, 500 * ONE_USDC],
+        charlie(),
+    )
+    .expect("Should successfully add liquidity");
+
+    assert_eq!(
+        share_price_and_total_shares(&mut session, stable_swap, None),
+        (last_share_price, last_total_shares + 1500 * ONE_LPT)
+    );
+
+    let last_total_shares = last_total_shares + 1500 * ONE_LPT;
+
+    // remove by shares (charlie)
+    _ = stable_swap::remove_liquidity_by_shares(
+        &mut session,
+        stable_swap,
+        CHARLIE,
+        300 * ONE_LPT,
+        vec![1 * ONE_DAI, 1 * ONE_USDT, 1 * ONE_USDC],
+        charlie(),
+    )
+    .expect("Should successfully remove liquidity");
+
+    assert_eq!(
+        psp22_utils::balance_of(&mut session, stable_swap, charlie()),
+        1200 * ONE_LPT
+    );
+    let balances: Vec<u128> = tokens
+        .iter()
+        .map(|&token| psp22_utils::balance_of(&mut session, token, charlie()))
+        .collect();
+    assert_eq!(
+        balances,
+        vec![100 * ONE_DAI, 100 * ONE_USDT, 100 * ONE_USDC],
+        "Incorrect Users tokens balances"
+    );
+    assert_eq!(
+        share_price_and_total_shares(&mut session, stable_swap, None),
+        (last_share_price, last_total_shares - 300 * ONE_LPT)
+    );
+    let last_total_shares = last_total_shares - 300 * ONE_LPT;
+
+    transfer_and_increase_allowance(
+        &mut session,
+        stable_swap,
+        tokens.clone(),
+        DAVE,
+        vec![100 * ONE_DAI, 200 * ONE_USDT, 400 * ONE_USDC],
+        BOB,
+    );
+
+    // add more liquidity with imbalanced tokens (dave)
+    _ = stable_swap::add_liquidity(
+        &mut session,
+        stable_swap,
+        DAVE,
+        1,
+        vec![100 * ONE_DAI, 200 * ONE_USDT, 400 * ONE_USDC],
+        dave(),
+    )
+    .expect("Should successfully add liquidity");
+    // "Mint 699699997426210330025 shares for user2, fee is 299999998348895348 shares",
+    // "Exchange swap got 59999999669779069 shares",
+
+    assert_eq!(
+        stable_swap::reserves(&mut session, stable_swap),
+        vec![100500 * ONE_DAI, 100600 * ONE_USDT, 100800 * ONE_USDC],
+        "Incorrect reserves"
+    );
+    assert_eq!(
+        psp22_utils::total_supply(&mut session, stable_swap),
+        301200 * ONE_LPT + 699699997426210330025 + 59999999669779069,
+        "Incorrect total shares"
+    );
+    assert_eq!(
+        psp22_utils::balance_of(&mut session, stable_swap, dave()),
+        699699997426210330025,
+        "Incorrect Users share"
+    );
+
+    let (current_share_price, current_total_shares) =
+        share_price_and_total_shares(&mut session, stable_swap, None);
+    assert!(
+        current_share_price > last_share_price,
```

Comment:
Does the share price increase when you add tokens in imbalanced amounts?

Reply:
Yes, b/c when depositing in imbalanced amounts then fees apply, hence the value of one share in terms of tokens in the pool increases.